### ISSUE TITLE
Automated build/deploy

### ISSUE THREAD

**Issue by Ajedi32** [metalsmith/metalsmith.io#149]
Would be nice if we didn't have to manually rebuild and deploy the site every time a change is made. Seems like this should be possible to do with Travis: https://docs.travis-ci.com/user/deployment/custom/

**Comment by Ajedi32** [comment]
Or, wait. Is this deployed on Heroku? In that case it's even easier I think: https://docs.travis-ci.com/user/deployment/heroku

**Comment by lambtron** [comment]
Yes, this is deployed on Heroku. I'll take a look at this!

**Comment by Ajedi32** [comment]
@lambtron Can you take a look at this when you get the chance? The current deploy system [still isn't working](https://github.com/metalsmith/metalsmith.io/pull/152#issuecomment-172291526).

**Comment by ismay** [comment]
Yeah this would be nice. There are currently still some plugins in `src/plugins.json` that aren't on the site yet.

**Comment by ismay** [comment]
How's this going? Is the build process still broken? Because I've added some plugins a while ago (https://github.com/metalsmith/metalsmith.io/pull/164) that still aren't on the site..

**Comment by fephil** [comment]
I have experience of using Travis CI to deploy to GitHub pages. I'd be happy to help do this if alternative deployment was being considered.

**Comment by Ajedi32** [comment]
@fephil: As @lambtron said, currently this site is deployed on Heroku. Given that it's all static content I see no reason why it *couldn't* be deployed on GitHub Pages, but you'd probably want to talk to @lambtron or someone else from Segment first to find out why Heroku was chosen as the deploy platform in the first place.

**Comment by lambtron** [comment]
Good idea moving this to github pages, i think that would be easier to manage. I'll take a look at this sometime this weekend.

cc @Ajedi32 @fephil

**Comment by Ajedi32** [comment]
Thanks. If I remember correctly, I think it's as simple as creating a `metalsmith.github.io` repo under the metalsmith org, then setting an ALIAS record in the DNS for `metalsmith.io` to point to `metalsmith.github.io`. After that, automating the deploy then becomes something we should be able to handle ourselves just fine, since we'd already have access to the `metalsmith.github.io` repo. (And here's some pretty good instructions for handling that: https://gist.github.com/domenic/ec8b0fc8ab45f39403dd)

**Comment by Ajedi32** [comment]
@lambtron I just got a new release of the site deployed to http://metalsmith.github.io/.

If you're too busy to work on this yourself, you can just create a DNS ALIAS record for metalsmith.io and point it at "metalsmith.github.io". We can probably take it from there...

One thing I did notice is that the fonts look a bit screwed up in the newly deployed site. I'm guessing that's because you're using proprietary (paid) fonts hosted on a third party site? Everything looked fine when I was running the server on localhost, but after deploying I'm getting a 403 Forbidden from https://cloud.typography.com/ while trying to fetch the fonts. I'm assuming that problem will be resolved when accessing the site from the correct domain (so cloud.typography.com will get a referrer header from a site authorized to use those paid fonts)?

**Comment by lambtron** [comment]
@ajedi32 Thank you so much for your contributions here; really appreciate it! I unfortunately don't have access to the DNS records myself, but I will get that ALIAS record updated.

I think you're assumption is correct on the font stuff. Let's try and find out.

**Comment by woodyrew** [comment]
@Ajedi32  and @lambtron Are we just waiting on the DNS ALIAS?

If not, Netlify is pretty good, and is [free for OSS projects](https://www.netlify.com/blog/2016/07/28/netlifys-pro-plan-now-free-for-open-source-projects/). Let me know and I'll try and have a go over the weekend.

**Comment by Ajedi32** [comment]
No, I haven't automated the deployment yet. I do have some changes though which still need to be merged. Specifically, #186. I had to revert it because it broke the real site currently hosted on Heroku.

**Comment by ismay** [comment]
Great to see some progress on this, awesome! Is there anything I can do to help? I'd like to get this up and running as it would be nice to have a plugins list that's up-to-date.

I've automated my deploys before with travis and github pages, it's fairly straightforward. I could take a look at that, or anything else. Just let me know where help is needed!

**Comment by woodyrew** [comment]
@Ajedi32 Is this on hold because of the GH pages deployment? It would be excellent if we could move forward, do you think @ismay can help you?

**Comment by Ajedi32** [comment]
@woodyrew Have a look at #193. I think that should solve this issue without needing to switch to GH pages (since apparently getting the DNS alias switched is a non-trivial task).

**Comment by Ajedi32** [comment]
@woodyrew Have a look at #193. I think that should solve this issue without needing to switch to GH pages (since apparently getting the DNS alias switched is a non-trivial task).

**Comment by ismay** [comment]
> Have a look at #193. I think that should solve this issue without needing to switch to GH pages

Ah great! Is it ready for merging or are there still things that need to be done?

> since apparently getting the DNS alias switched is a non-trivial task

Just wondering, but shouldn't this be relatively straightforward, as long as the person who has the credentials just points it somewhere else right?

Seems like merging #193 and then discussing if and how to transition to gh-pages would be a good way to move forward. I'd personally be in favor of still moving to github pages as it would be much easier to manage. All our interaction happens on here and the site being served by github as well would simplify the whole process in my opinion. But I'm curious to hear what everybody else thinks about that.

**Comment by Ajedi32** [comment]
> Is it ready for merging or are there still things that need to be done?

Should be good to merge. There are other things which can be done, since it's not using gh-pages yet (or Netlify, or whatever), but that PR should at least solve this issue by automatically deploying to Heroku.

**Comment by Ajedi32** [comment]
> > since apparently getting the DNS alias switched is a non-trivial task
>
> Just wondering, but shouldn't this be relatively straightforward, as long as the person who has the credentials just points it somewhere else right?

It _should_ be straightforward, but apparently nobody who works on Metalsmith actually has access to the controls for the metalsmith.io domain. AFAIK the domain is owned by [Segment](https://segment.com/), and while several of the Metalsmith org's members are Segment employees, even they don't have direct access to change DNS records for Metalsmith. (See @lambtron's [comment](https://github.com/metalsmith/metalsmith.io/issues/149#issuecomment-242943439) above.)

**Comment by Ajedi32** [comment]
> I'd personally be in favor of still moving to github pages as it would be much easier to manage. All our interaction happens on here and the site being served by github as well would simplify the whole process in my opinion.

If deploys are automatic, I'm actually not sure it makes much of a difference what host we use as far as process goes. I do agree though GH Pages is probably better than Heroku, mainly because of simplicity. Since the Metalsmith site is currently just static pages, Heroku is somewhat overkill.

The downside to GH Pages is that it currently doesn't directly support HTTPS on custom domains, and I'd like to get HTTPS support up and running if possible (#187). Maybe we could point the domain at CloudFlare and manage HTTPS that way...

**Comment by ismay** [comment]
> Should be good to merge. There are other things which can be done, since it's not using gh-pages yet (or Netlify, or whatever), but that PR should at least solve this issue by automatically deploying to Heroku.

Ah that's great! That would at least fix things for now while we work on the dns/ns stuff.

> It should be straightforward, but apparently nobody who works on Metalsmith actually has access to the controls for the metalsmith.io domain. AFAIK the domain is owned by Segment, and while several of the Metalsmith org's members are Segment employees, even they don't have direct access to change DNS records for Metalsmith.

Hmm ok. Hopefully this can be picked up by someone at segment then, otherwise we won't even have the option of moving to gh-pages or something else.

> If deploys are automatic, I'm actually not sure it makes much of a difference what host we use as far as process goes. I do agree though GH Pages is probably better than Heroku, mainly because of simplicity. Since the Metalsmith site is currently just static pages, Heroku is somewhat overkill.

Definitely, seems overly complicated for a static site. But yeah, in practice the result is the same whether you're hosting on heroku or gh, just makes maintenance/deployment less simple.

> The downside to GH Pages is that it currently doesn't directly support HTTPS on custom domains, and I'd like to get HTTPS support up and running if possible (#187). Maybe we could point the domain at CloudFlare and manage HTTPS that way...

That would be my solution too, I do that as well. So instead of updating the alias we should then point the NS for metalsmith.io to cloudflare. If that's not an option, maybe https://surge.sh is a good alternative.

**Comment by woodyrew** [comment]
I'll get this merged and then we can move forwards with some of the other issues/PRs

I've spoken to lambtron about the DNS, he's contacted contacted the CTO and linked to this issue so hopefully we'll make some progress soon.

**Comment by woodyrew** [comment]
Changes are automatically deploying to metalsmith.io. with #193 (Thanks @Ajedi32).

The only concern I have is that there's no feedback regarding the build or the deployment at the moment. Is there a status for this anywhere with Heroku in the short term?

**Comment by woodyrew** [comment]
I've added an issue to [add verification tests](#195).

Also, reopening this issue as I think it would be good to move over to GH pages (if we can have https easily), Netlifiy or Surge.

**Comment by Ajedi32** [comment]
Wouldn't it be better to open another issue for that? While there has been a lot of talk here about switching to GitHub pages or something similar, this issue is still about automating the deploy, and it appears that particular issue is now solved.

**Comment by woodyrew** [comment]
I've opened #197 and #198 to discuss the outstanding issues.
